### PR TITLE
Bound Unsigned when signed is not supported.

### PR DESCRIPTION
**Pull request by thiolliere**
Fix https://github.com/paritytech/substrate/issues/3716

### Context

Perbill and other perthings do not work with signed arithmetic: `Percent(50) * -12i32 == 127 != -6` though they allow it.
Some generic bound AtLeast32Bit but actually doesn't support signed type (e.g. `pallet-balance::Trait::Balance`)

### Done:

* created: AtLeast32BitUnsigned
* make perthings bound unsigned when needed
* make sp-runtime/curve use unsigned only
* make balance use unsigned only
* make block number use unsigned only
  some generic function seems to handle signed block number, some doesn't, I decided to use `Unsigned` anyway in all use of block number.
* make pallet-asset balance use unsigned (I'm not sure the logic is sound if we accept signed balance, e.g for transfer)
* make Header number being unsigned (this bound doesn't seem strictly needed right now, maybe I can keep allowing signed ones ?)

### Todo:

* [ ] might need a polkadot PR

### Additional notes

* I'm not sure about Header number in primitives, should it allow signed ones ?
* We could change AtLeast32Bit to bound Unsigned directly.
* I can't ensure that there is no missed generic function which incorrectly allow signed integer. But AtLeast32Bit is only used in:
  * `frame-system::Trait::Index`
  * `generic-asset::Trait::AssetId`
  * `scored-pool::Trait::Score`
  * `asset::Trait::AssetId`
  * in some parts of client that I can't really tell.

### CODE REVIEW

File: primitives/arithmetic/src/traits.rs [hunk 1]
```diff
@@ -79,6 +79,15 @@ pub trait AtLeast32Bit: BaseArithmetic + From<u16> + From<u32> {}
 
 impl<T: BaseArithmetic + From<u16> + From<u32>> AtLeast32Bit for T {}
 
+/// A meta trait for arithmetic.
+///
+/// Arithmetic types do all the usual stuff you'd expect numbers to do. They are guaranteed to
+/// be able to represent at least `u32` values without loss, hence the trait implies `From<u32>`
+/// and smaller integers. All other conversions are fallible. And is unsigned.
```
Comment (kianenigma) [inline primitives/arithmetic/src/traits.rs:86]: I'd be fine here by just saying: 

```suggestion
/// Same as [`AtLeast32Bit `], but also bounded to be unsigned. 
```

File: frame/generic-asset/src/lib.rs [hunk 2]
```diff
@@ -178,25 +178,15 @@ mod tests;
...
 
 pub trait Trait: frame_system::Trait {
-	type Balance: Parameter
-		+ Member
-		+ AtLeast32Bit
-		+ Default
-		+ Copy
-		+ MaybeSerializeDeserialize
-		+ Debug;
+	type Balance: Parameter + Member + AtLeast32BitUnsigned + Default + Copy +
```
Comment (kianenigma) [inline frame/generic-asset/src/lib.rs:181]: can probably fit even in one line?

File: primitives/arithmetic/src/per_things.rs [hunk 3]
```diff
@@ -330,14 +334,15 @@ macro_rules! implement_per_thing {
 			/// Build this type from a number of parts per thing.
 			fn from_parts(parts: Self::Inner) -> Self { Self(parts.min($max)) }
 
+			/// NOTE: saturate to 0 or 1 if x is beyond `[0, 1]`
 			#[cfg(feature = "std")]
 			fn from_fraction(x: f64) -> Self {
-				Self::from_parts((x * $max as f64) as Self::Inner)
+				Self::from_parts((x.max(0.).min(1.) * $max as f64) as Self::Inner)
```
Comment (kianenigma) [inline primitives/arithmetic/src/per_things.rs:340]: 👍🏻

**Comment by kianenigma** [comment]
> make pallet-asset balance use unsigned (I'm not sure the logic is sound if we accept signed balance, e.g for transfer)

I think balance should only work with unsigned. 

>  might need a polkadot PR

Seemingly not.


> I'm not sure about Header number in primitives, should it allow signed ones ?

Not sure but probably worth asking around. 

> We could change AtLeast32Bit to bound Unsigned directly.

I prefer having two variants: `AtLeast32Bit` and `AtLeast32BitUnsigned`.

File: frame/generic-asset/src/lib.rs (hunk 2 above)
Comment (thiolliere) [reply frame/generic-asset/src/lib.rs:181]: not in 100 character but fits in 120, anyway I just reorganize slightly

**Pull request merged**